File: app/src/main/java/de/hdc/kspchecklist/ItemListAdapter.java
package de.hdc.kspchecklist;

import android.content.Context;
import android.support.annotation.NonNull;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.TextView;

import java.util.ArrayList;

import de.hdc.kspchecklist.data.ListItem;

/**
 * Created by DerTroglodyt on 2017-02-16 08:53.
 * Email dertroglodyt@gmail.com
 * Copyright by HDC, Germany
 */

public class ItemListAdapter extends ArrayAdapter<ListItem> {

    public ItemListAdapter(Context context, ArrayList<ListItem> objects) {
        super(context, 0, objects);
    }

    @NonNull
    @Override
    public View getView(int position, View convertView, ViewGroup parent) {
        final ListItem item = getItem(position);
        ViewHolder viewHolder; // view lookup cache stored in tag

        if (convertView == null) {
            // If there's no view to re-use, inflate a brand new view for row
            viewHolder = new ViewHolder();
            LayoutInflater inflater = LayoutInflater.from(getContext());
            convertView = inflater.inflate(R.layout.item_list_content, parent, false);
            viewHolder.name = (TextView) convertView.findViewById(R.id.content);
            // Cache the viewHolder object inside the fresh view
            convertView.setTag(viewHolder);
        } else {
            // View is being recycled, retrieve the viewHolder object from tag
            viewHolder = (ViewHolder) convertView.getTag();
        }
        // Populate the data from the data object via the viewHolder object
        // into the template view.
        viewHolder.name.setText(item.name);
        // Return the completed view to render on screen
        return convertView;
    }

    // View lookup cache
    private static class ViewHolder {
        TextView name;
    }
}

File: app/src/main/java/de/hdc/kspchecklist/ItemListAdapter.kt
package de.hdc.kspchecklist

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ArrayAdapter
import android.widget.TextView
import de.hdc.kspchecklist.data.ListItem
import java.util.*

/**
 * Created by DerTroglodyt on 2017-02-16 08:53.
 * Email dertroglodyt@gmail.com
 * Copyright by HDC, Germany
 */

internal class ItemListAdapter(context: Context, objects: ArrayList<ListItem>)
    : ArrayAdapter<ListItem>(context, 0, objects) {

    override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
        val view: View
        val viewHolder: ViewHolder // view lookup cache stored in tag

        if (convertView == null) {
            // If there's no view to re-use, inflate a brand new view for row
            val inflater = LayoutInflater.from(context)
            view = inflater.inflate(R.layout.item_list_content, parent, false)
            viewHolder = ViewHolder(view.findViewById<View>(R.id.content) as TextView)
            // Cache the viewHolder object inside the fresh view
            view.tag = viewHolder
        } else {
            // View is being recycled, retrieve the viewHolder object from tag
            view = convertView
            viewHolder = view.tag as ViewHolder
        }
        // Populate the data from the data object via the viewHolder object
        // into the template view.
        viewHolder.name.text = getItem(position).name

        return view
    }

    // View lookup cache
    private class ViewHolder(val name: TextView)
}
